packages/shorebird_cli/lib/src/commands/patch/macos_patcher.dart

```diff
@@ -0,0 +1,446 @@
+          codesign: codesign,
+          flavor: flavor,
+          target: target,
+          args: argResults.forwardedArgs +
+              buildNameAndNumberArgsFromReleaseVersion(releaseVersion),
+          base64PublicKey: argResults.encodedPublicKey,
+          buildProgress: buildProgress,
+        );
+      } on ProcessException catch (error) {
+        buildProgress.fail('Failed to build: ${error.message}');
+        rethrow;
+      } on ArtifactBuildException catch (error) {
+        buildProgress.fail('Failed to build macOS app');
+        logger.err(error.message);
+        rethrow;
+      }
+
+      try {
+        if (splitDebugInfoPath != null) {
+          Directory(splitDebugInfoPath!).createSync(recursive: true);
+        }
+        await artifactBuilder.buildElfAotSnapshot(
+          appDillPath: macosBuildResult.kernelFile.path,
+          outFilePath: _aotOutputPath,
+          genSnapshotArtifact: ShorebirdArtifact.genSnapshotMacOS,
+          additionalArgs: splitDebugInfoArgs(splitDebugInfoPath),
+        );
+      } catch (error) {
+        buildProgress.fail('$error');
+        rethrow;
+      }
+
+      // Copy the kernel file to the build directory so that it can be used
+      // to generate a patch.
+      macosBuildResult.kernelFile.copySync(_appDillCopyPath);
+
+      buildProgress.complete();
+    } catch (_) {
+      throw ProcessExit(ExitCode.software.code);
+    }
+
+    final appPath = artifactManager.getMacOSAppDirectory()!.path;
+    final tempDir = await Directory.systemTemp.createTemp();
+    final zippedApp = File(p.join(tempDir.path, '${p.basename(appPath)}.zip'));
+    await ditto.archive(
+      source: appPath,
+      destination: zippedApp.path,
+      keepParent: true,
+    );
+    return zippedApp;
+  }
+
+  @override
+  Future<Map<Arch, PatchArtifactBundle>> createPatchArtifacts({
+    required String appId,
+    required int releaseId,
+    required File releaseArtifact,
+    File? supplementArtifact,
+  }) async {
+    // Verify that we have built a patch .app
+    if (artifactManager.getMacOSAppDirectory()?.path == null) {
+      logger.err('Unable to find .app directory');
+      throw ProcessExit(ExitCode.software.code);
+    }
+
+    final unzipProgress = logger.progress('Extracting release artifact');
+    final releaseAppDirectory = Directory.systemTemp.createTempSync();
+    await ditto.extract(
+      source: releaseArtifact.path,
+      destination: releaseAppDirectory.path,
+    );
+
+    File? releaseClassTableLinkInfoFile;
+    File? releaseClassTableLinkDebugInfoFile;
+    if (supplementArtifact != null) {
+      final tempDir = Directory.systemTemp.createTempSync();
+      await artifactManager.extractZip(
+        zipFile: supplementArtifact,
+        outputDirectory: tempDir,
+      );
+      releaseClassTableLinkInfoFile = File(p.join(tempDir.path, 'App.ct.link'));
+      if (!releaseClassTableLinkInfoFile.existsSync()) {
```

Comment:
I probably would have had an _existsOrExit(file, 'class table link debug info') if I'm doing this more often. 🤷

Reply:
I'm planning to refactor all this and extract the shared logic in a subsequent PR.